README.md
# ApplyAI (apply-ai.online)

[Live Version](https://apply-ai.online)

[Visit Devpost Page](https://devpost.com/software/applyai)

[Demo Video](https://www.youtube.com/watch?v=7Sxl3rX8u4U)

[Front End Repo](https://github.com/nicholas-tao/apply-ai-frontend)

[Parsers](https://github.com/qasimza/apply-ai-parsers)

## Inspiration

As a result of the COVID-19 pandemic, millions of people lost their jobs and the US unemployment rate reached an all-time-high of nearly 15% in April of 2020. Job hunting has never been easy. It's time consuming, repetitive, and draining.

As students, we've spent countless hours applying to internships and filling in the same details over and over again.

ApplyAI is here to change that.

## What it does

ApplyAI is broken into two main features:

First, it parses your resume to automatically fill out as many fields as it can (if you wrote it on your resume once, why should you have to also enter it into a form a million times?). It then will let you modify any of the fields it gathered, and then update your information.

Second, our extensive bot network will be constantly adding new jobs to our database of jobs and will automatically pick a few that best suit you, using fuzzy matching and cosine similarity. These jobs will be displayed to you all in one place at one time, and you can select as many as you would like and click apply once. That's it! You just applied to multiple highly-relevant jobs in under a minute!

## How we built it

The frontend is built with HTML, CSS, JS, and Bootstrap. We use Javascript's Fetch API to communicate with the backend. The frontend is hosted on Firebase at apply-ai.online.

The backend is a Ubuntu server configured with Nginx serving files to api.apply-ai.online. It is built using python and flask, which then communicates with a Firebase Realtime Database and a Google Cloud Storage Bucket to store resumes.

This project also has a cronjob running on our Ubuntu server. This cronjob is constantly indexing many job sites to pull as much useful info as it can, and continues to train our ML model as it goes. This ensures that even with a limited number of jobs on the market, our algorithms will always be able to provide you with the best job recommendations out there.

## Challenges we ran into

Connecting the frontend to the backend took a long time. We, rather naively, began connecting the two at about 2 am on the day of submissions. Integrating the two was not successful until 10 am on the day of submissions. This meant our entire Devpost is being written in an extremely sleep-deprived state, so we hope you like it!

Besides the constant lack of sleep, another challenge was parsing the resume. Unfortunately, pdf resumes don't have a standardized format. This meant in many cases we needed to apply ML instead of an algorithm. Even still, these ML models are by no means perfect. If you run our demo and happen to get weird data in your pre-filled forms, or no data at all, that is likely a result of our model being unable to properly scrape your resume. If you would be willing, we would love to receive your resume so we could work on upgrading the model in the future to support it.

## Accomplishments that we're proud of

With only 3 members on our team, and none of us heavy in frontend knowledge, we knew this was an ambitious project. The fact that we got the Devpost submitted is a testament to our perseverance as a team, especially considering I (Eric), only slept about 6 hours this entire hackathon.

## What we learned

As cheesy as it sounds, I really learned to be patient. This took a lot of time and restraint while up in the night debugging. When we finally made it through with everything working, it was the greatest feeling of relief ever.

In addition to our excellent practice debugging it, we also learned how to integrate so many moving parts. We wanted to implement many different features so we would be in the running for the Google Cloud prize track. This meant if one thing threw an error we weren't catching, everything would break. We learned just how important proper error handling is.

## What's next for ApplyAI

We really hope to improve on our resume parser. Currently, that is definitely a bottleneck.

We also want to integrate our bots with more job posting websites, like monster.com.

Lastly, our bots have trouble applying to many websites where you need to create an account, or there is an "I'm not a robot" functionality. We are thinking of perhaps packaging this as a chrome extension, and when it detects job fields it will fill them out for you.


changelog.txt
Version: 2.2.0
  - Updated the PHP Email Form to v2.1
  - Other small fixes and improvements
  
Version: 2.1.0
  - Updated Bootstrap to version 4.5.0
  - Updated the PHP Email Form library to version 2.0 with reCaptcha support
  - Aded inner-page.html tempalte
  - Added smooth scroll on page load with hash links in the url
  - Updated all outdated third party vendor libraries to their latest versions
  - Other small fixes and improvements
  
Version: 2.0.0
  - The template was rebuilt from scratch with the latest Bootstrap version (4.4.1)
  - Added SMPTP support for the contact form script (Pro)
  - Added NodeJS NPM Development version (Pro unlimited & Membership members)
  
Version: 1.0.0
  - Initial Release


src/components/sections/Hero.js
import React, { useState } from "react";
import classNames from "classnames";
import { SectionProps } from "../../utils/SectionProps";
import ButtonGroup from "../elements/ButtonGroup";
import Button2 from "../elements/Button";
import Image from "../elements/Image";
import Modal from "../elements/Modal";
import Form from "react-bootstrap/Form";
import Button from "react-bootstrap/Button";

const propTypes = {
  ...SectionProps.types,
};

const defaultProps = {
  ...SectionProps.defaults,
};

const Hero = ({
  className,
  topOuterDivider,
  bottomOuterDivider,
  topDivider,
  bottomDivider,
  hasBgColor,
  invertColor,
  ...props
}) => {
  const [videoModalActive, setVideomodalactive] = useState(false);

  const openModal = (e) => {
    e.preventDefault();
    setVideomodalactive(true);
  };

  const closeModal = (e) => {
    e.preventDefault();
    setVideomodalactive(false);
  };

  const outerClasses = classNames(
    "hero section center-content",
    topOuterDivider && "has-top-divider",
    bottomOuterDivider && "has-bottom-divider",
    hasBgColor && "has-bg-color",
    invertColor && "invert-color",
    className
  );

  const innerClasses = classNames(
    "hero-inner section-inner",
    topDivider && "has-top-divider",
    bottomDivider && "has-bottom-divider"
  );

  return (
    <section {...props} className={outerClasses}>
      <div className="container-sm">
        <div className={innerClasses}>
          <div className="hero-content">
            <h1
              className="mt-0 mb-16 reveal-from-bottom"
              data-reveal-delay="200"
            >
              Job hunting made <span className="text-color-primary">easy</span>
            </h1>
            <div className="container-xs">
              <p
                className="m-0 mb-32 reveal-from-bottom"
                data-reveal-delay="400"
              >
                We find and apply to jobs for you.
              </p>
              <div className="reveal-from-bottom" data-reveal-delay="600">
                <ButtonGroup>
                  <Button2
                    tag="a"
                    color="primary"
                    wideMobile
                    href="https://cruip.com/"
                  >
                    Get started
                  </Button2>
                  {/*
                     <Button
                    tag="a"
                    color="dark"
                    wideMobile
                    href="https://github.com/cruip/open-react-template/"
                  >
                    View on Github
                  </Button>
                    */}
                </ButtonGroup>
              </div>
            </div>
          </div>
          <div>
            <Form>
              <Form.Group controlId="formBasicEmail">
                <Form.Label>Email address</Form.Label>
                <Form.Control type="email" placeholder="Enter email" />
              </Form.Group>

              <Button variant="primary" type="submit">
                Submit
              </Button>
            </Form>
          </div>
          <div
            className="hero-figure reveal-from-bottom illustration-element-01"
            data-reveal-value="20px"
            data-reveal-delay="800"
          >
            {/*

              <a
              data-video="https://player.vimeo.com/video/174002812"
              href="#0"
              aria-controls="video-modal"
              onClick={openModal}
            >
              <Image
                className="has-shadow"
                src={require("./../../assets/images/video-placeholder.jpg")}
                alt="Hero"
                width={896}
                height={504}
              />
            </a>
              */}
          </div>
          <Modal
            id="video-modal"
            show={videoModalActive}
            handleClose={closeModal}
            video="https://player.vimeo.com/video/174002812"
            videoTag="iframe"
          />
        </div>
      </div>
    </section>
  );
};

Hero.propTypes = propTypes;
Hero.defaultProps = defaultProps;

export default Hero;


public/assets/js/template.js
//template js file
!(function ($) {
  "use strict";

  // Smooth scroll for the navigation menu and links with .scrollto classes
  var scrolltoOffset = $("#header").outerHeight() - 15;
  $(document).on("click", ".nav-menu a, .mobile-nav a, .scrollto", function (
    e
  ) {
    if (
      location.pathname.replace(/^\//, "") ==
        this.pathname.replace(/^\//, "") &&
      location.hostname == this.hostname
    ) {
      var target = $(this.hash);
      if (target.length) {
        e.preventDefault();

        var scrollto = target.offset().top - scrolltoOffset;

        if ($(this).attr("href") == "#header") {
          scrollto = 0;
        }

        $("html, body").animate(
          {
            scrollTop: scrollto,
          },
          1500,
          "easeInOutExpo"
        );

        if ($(this).parents(".nav-menu, .mobile-nav").length) {
          $(".nav-menu .active, .mobile-nav .active").removeClass("active");
          $(this).closest("li").addClass("active");
        }

        if ($("body").hasClass("mobile-nav-active")) {
          $("body").removeClass("mobile-nav-active");
          $(".mobile-nav-toggle i").toggleClass(
            "icofont-navigation-menu icofont-close"
          );
          $(".mobile-nav-overly").fadeOut();
        }
        return false;
      }
    }
  });

  // Activate smooth scroll on page load with hash links in the url
  $(document).ready(function () {
    if (window.location.hash) {
      var initial_nav = window.location.hash;
      if ($(initial_nav).length) {
        var scrollto = $(initial_nav).offset().top - scrolltoOffset;
        $("html, body").animate(
          {
            scrollTop: scrollto,
          },
          1500,
          "easeInOutExpo"
        );
      }
    }
  });

  // Mobile Navigation
  if ($(".nav-menu").length) {
    var $mobile_nav = $(".nav-menu").clone().prop({
      class: "mobile-nav d-lg-none",
    });
    $("body").append($mobile_nav);
    $("body").prepend(
      '<button type="button" class="mobile-nav-toggle d-lg-none"><i class="icofont-navigation-menu"></i></button>'
    );
    $("body").append('<div class="mobile-nav-overly"></div>');

    $(document).on("click", ".mobile-nav-toggle", function (e) {
      $("body").toggleClass("mobile-nav-active");
      $(".mobile-nav-toggle i").toggleClass(
        "icofont-navigation-menu icofont-close"
      );
      $(".mobile-nav-overly").toggle();
    });

    $(document).on("click", ".mobile-nav .drop-down > a", function (e) {
      e.preventDefault();
      $(this).next().slideToggle(300);
      $(this).parent().toggleClass("active");
    });

    $(document).click(function (e) {
      var container = $(".mobile-nav, .mobile-nav-toggle");
      if (!container.is(e.target) && container.has(e.target).length === 0) {
        if ($("body").hasClass("mobile-nav-active")) {
          $("body").removeClass("mobile-nav-active");
          $(".mobile-nav-toggle i").toggleClass(
            "icofont-navigation-menu icofont-close"
          );
          $(".mobile-nav-overly").fadeOut();
        }
      }
    });
  } else if ($(".mobile-nav, .mobile-nav-toggle").length) {
    $(".mobile-nav, .mobile-nav-toggle").hide();
  }

  // Navigation active state on scroll
  var nav_sections = $("section");
  var main_nav = $(".nav-menu, #mobile-nav");

  $(window).on("scroll", function () {
    var cur_pos = $(this).scrollTop() + 200;

    nav_sections.each(function () {
      var top = $(this).offset().top,
        bottom = top + $(this).outerHeight();

      if (cur_pos >= top && cur_pos <= bottom) {
        if (cur_pos <= bottom) {
          main_nav.find("li").removeClass("active");
        }
        main_nav
          .find('a[href="#' + $(this).attr("id") + '"]')
          .parent("li")
          .addClass("active");
      }
      if (cur_pos < 300) {
        $(".nav-menu ul:first li:first").addClass("active");
      }
    });
  });

  // Toggle .header-scrolled class to #header when page is scrolled
  $(window).scroll(function () {
    if ($(this).scrollTop() > 100) {
      $("#header").addClass("header-scrolled");
    } else {
      $("#header").removeClass("header-scrolled");
    }
  });

  if ($(window).scrollTop() > 100) {
    $("#header").addClass("header-scrolled");
  }

  // Back to top button
  $(window).scroll(function () {
    if ($(this).scrollTop() > 100) {
      $(".back-to-top").fadeIn("slow");
    } else {
      $(".back-to-top").fadeOut("slow");
    }
  });

  $(".back-to-top").click(function () {
    $("html, body").animate(
      {
        scrollTop: 0,
      },
      1500,
      "easeInOutExpo"
    );
    return false;
  });

  // jQuery counterUp
  $('[data-toggle="counter-up"]').counterUp({
    delay: 10,
    time: 1000,
  });

  // Testimonials carousel (uses the Owl Carousel library)
  $(".testimonials-carousel").owlCarousel({
    autoplay: true,
    dots: true,
    loop: true,
    responsive: {
      0: {
        items: 1,
      },
      768: {
        items: 1,
      },
      900: {
        items: 2,
      },
    },
  });

  // Porfolio isotope and filter
  $(window).on("load", function () {
    var portfolioIsotope = $(".portfolio-container").isotope({
      itemSelector: ".portfolio-item",
      layoutMode: "fitRows",
    });

    $("#portfolio-flters li").on("click", function () {
      $("#portfolio-flters li").removeClass("filter-active");
      $(this).addClass("filter-active");

      portfolioIsotope.isotope({
        filter: $(this).data("filter"),
      });
      aos_init();
    });

    // Initiate venobox (lightbox feature used in portofilo)
    $(document).ready(function () {
      $(".venobox").venobox();
    });
  });

  // Portfolio details carousel
  $(".portfolio-details-carousel").owlCarousel({
    autoplay: true,
    dots: true,
    loop: true,
    items: 1,
  });

  // Init AOS
  function aos_init() {
    AOS.init({
      duration: 1000,
      easing: "ease-in-out",
      once: true,
      mirror: false,
    });
  }
  $(window).on("load", function () {
    aos_init();
  });
})(jQuery);
